From ea84b5beb207b329435be1f294ff42f51742a05d Mon Sep 17 00:00:00 2001
From: Dan Prince <dprince@redhat.com>
Date: Mon, 8 Sep 2025 08:53:27 -0400
Subject: [PATCH 4/5] Change RabbitmqClusterSpecCore.Spec.Override.StatefulSet
 to raw json type

Use raw json instead of nesting a giant struct

Jira: OSPRH-17003
---
 api/v1beta1/rabbitmqcluster_types.go          | 10 +++-
 api/v1beta1/zz_generated.deepcopy.go          | 31 ++++++++++-
 .../bases/rabbitmq.com_rabbitmqclusters.yaml  | 51 ++++---------------
 docs/api/rabbitmq.com.ref.asciidoc            | 20 +++++++-
 4 files changed, 67 insertions(+), 45 deletions(-)

diff --git a/api/v1beta1/rabbitmqcluster_types.go b/api/v1beta1/rabbitmqcluster_types.go
index 53912bfc..baaa00eb 100644
--- a/api/v1beta1/rabbitmqcluster_types.go
+++ b/api/v1beta1/rabbitmqcluster_types.go
@@ -20,6 +20,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	k8sresource "k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 )
 
 const DisableDefaultTopologySpreadAnnotation = "rabbitmq.com/disable-default-topology-spread-constraints"
@@ -106,6 +107,13 @@ type RabbitmqClusterSpec struct {
 	SecretBackend SecretBackend `json:"secretBackend,omitempty"`
 }
 
+type OverrideTrimmed struct {
+	// Override configuration for the RabbitMQ StatefulSet.
+	StatefulSet *runtime.RawExtension `json:"statefulSet,omitempty"`
+	// Override configuration for the Service created to serve traffic to the cluster.
+	Service *Service `json:"service,omitempty"`
+}
+
 // A duplicate of the RabbitmqClusterSpec, but with the Image fields removed for use by OpenStackControlplane
 // NOTE: we duplicate it to keep the delta/PR here lighteweight and avoid changes to tests/controllers. This
 // will need to be recopied/created if the RabbitmqClusterSpec changes.
@@ -133,7 +141,7 @@ type RabbitmqClusterSpecCore struct {
 	// TLS-related configuration for the RabbitMQ cluster.
 	TLS TLSSpec `json:"tls,omitempty"`
 	// Provides the ability to override the generated manifest of several child resources.
-	Override RabbitmqClusterOverrideSpec `json:"override,omitempty"`
+	Override *OverrideTrimmed `json:"override,omitempty"`
 	// If unset, or set to false, the cluster will run `rabbitmq-queues rebalance all` whenever the cluster is updated.
 	// Set to true to prevent the operator rebalancing queue leaders after a cluster update.
 	// Has no effect if the cluster only consists of one node.
diff --git a/api/v1beta1/zz_generated.deepcopy.go b/api/v1beta1/zz_generated.deepcopy.go
index 819edc41..2207c0ad 100644
--- a/api/v1beta1/zz_generated.deepcopy.go
+++ b/api/v1beta1/zz_generated.deepcopy.go
@@ -80,6 +80,31 @@ func (in *EmbeddedObjectMeta) DeepCopy() *EmbeddedObjectMeta {
 	return out
 }
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (in *OverrideTrimmed) DeepCopyInto(out *OverrideTrimmed) {
+	*out = *in
+	if in.StatefulSet != nil {
+		in, out := &in.StatefulSet, &out.StatefulSet
+		*out = new(runtime.RawExtension)
+		(*in).DeepCopyInto(*out)
+	}
+	if in.Service != nil {
+		in, out := &in.Service, &out.Service
+		*out = new(Service)
+		(*in).DeepCopyInto(*out)
+	}
+}
+
+// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new OverrideTrimmed.
+func (in *OverrideTrimmed) DeepCopy() *OverrideTrimmed {
+	if in == nil {
+		return nil
+	}
+	out := new(OverrideTrimmed)
+	in.DeepCopyInto(out)
+	return out
+}
+
 // DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
 func (in *PersistentVolumeClaim) DeepCopyInto(out *PersistentVolumeClaim) {
 	*out = *in
@@ -435,7 +460,11 @@ func (in *RabbitmqClusterSpecCore) DeepCopyInto(out *RabbitmqClusterSpecCore) {
 	}
 	in.Rabbitmq.DeepCopyInto(&out.Rabbitmq)
 	out.TLS = in.TLS
-	in.Override.DeepCopyInto(&out.Override)
+	if in.Override != nil {
+		in, out := &in.Override, &out.Override
+		*out = new(OverrideTrimmed)
+		(*in).DeepCopyInto(*out)
+	}
 	if in.TerminationGracePeriodSeconds != nil {
 		in, out := &in.TerminationGracePeriodSeconds, &out.TerminationGracePeriodSeconds
 		*out = new(int64)
diff --git a/config/crd/bases/rabbitmq.com_rabbitmqclusters.yaml b/config/crd/bases/rabbitmq.com_rabbitmqclusters.yaml
index a0b334a5..ca1b68b2 100644
--- a/config/crd/bases/rabbitmq.com_rabbitmqclusters.yaml
+++ b/config/crd/bases/rabbitmq.com_rabbitmqclusters.yaml
@@ -10,7 +10,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.18.0
+    controller-gen.kubebuilder.io/version: v0.16.3
   name: rabbitmqclusters.rabbitmq.com
 spec:
   group: rabbitmq.com
@@ -329,6 +329,7 @@ spec:
                                       pod labels will be ignored. The default value is empty.
                                       The same key is forbidden to exist in both matchLabelKeys and labelSelector.
                                       Also, matchLabelKeys cannot be set when labelSelector isn't set.
+                                      This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                     items:
                                       type: string
                                     type: array
@@ -343,6 +344,7 @@ spec:
                                       pod labels will be ignored. The default value is empty.
                                       The same key is forbidden to exist in both mismatchLabelKeys and labelSelector.
                                       Also, mismatchLabelKeys cannot be set when labelSelector isn't set.
+                                      This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                     items:
                                       type: string
                                     type: array
@@ -503,6 +505,7 @@ spec:
                                   pod labels will be ignored. The default value is empty.
                                   The same key is forbidden to exist in both matchLabelKeys and labelSelector.
                                   Also, matchLabelKeys cannot be set when labelSelector isn't set.
+                                  This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                 items:
                                   type: string
                                 type: array
@@ -517,6 +520,7 @@ spec:
                                   pod labels will be ignored. The default value is empty.
                                   The same key is forbidden to exist in both mismatchLabelKeys and labelSelector.
                                   Also, mismatchLabelKeys cannot be set when labelSelector isn't set.
+                                  This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                 items:
                                   type: string
                                 type: array
@@ -670,6 +674,7 @@ spec:
                                       pod labels will be ignored. The default value is empty.
                                       The same key is forbidden to exist in both matchLabelKeys and labelSelector.
                                       Also, matchLabelKeys cannot be set when labelSelector isn't set.
+                                      This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                     items:
                                       type: string
                                     type: array
@@ -684,6 +689,7 @@ spec:
                                       pod labels will be ignored. The default value is empty.
                                       The same key is forbidden to exist in both mismatchLabelKeys and labelSelector.
                                       Also, mismatchLabelKeys cannot be set when labelSelector isn't set.
+                                      This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                     items:
                                       type: string
                                     type: array
@@ -844,6 +850,7 @@ spec:
                                   pod labels will be ignored. The default value is empty.
                                   The same key is forbidden to exist in both matchLabelKeys and labelSelector.
                                   Also, matchLabelKeys cannot be set when labelSelector isn't set.
+                                  This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                 items:
                                   type: string
                                 type: array
@@ -858,6 +865,7 @@ spec:
                                   pod labels will be ignored. The default value is empty.
                                   The same key is forbidden to exist in both mismatchLabelKeys and labelSelector.
                                   Also, mismatchLabelKeys cannot be set when labelSelector isn't set.
+                                  This is a beta field and requires enabling MatchLabelKeysInPodAffinity feature gate (enabled by default).
                                 items:
                                   type: string
                                 type: array
@@ -1842,8 +1850,6 @@ spec:
                                                       - port
                                                     type: object
                                                 type: object
-                                              stopSignal:
-                                                type: string
                                             type: object
                                           livenessProbe:
                                             properties:
@@ -2554,8 +2560,6 @@ spec:
                                                       - port
                                                     type: object
                                                 type: object
-                                              stopSignal:
-                                                type: string
                                             type: object
                                           livenessProbe:
                                             properties:
@@ -3280,8 +3284,6 @@ spec:
                                                       - port
                                                     type: object
                                                 type: object
-                                              stopSignal:
-                                                type: string
                                             type: object
                                           livenessProbe:
                                             properties:
@@ -3786,39 +3788,6 @@ spec:
                                       x-kubernetes-list-map-keys:
                                         - name
                                       x-kubernetes-list-type: map
-                                    resources:
-                                      properties:
-                                        claims:
-                                          items:
-                                            properties:
-                                              name:
-                                                type: string
-                                              request:
-                                                type: string
-                                            required:
-                                              - name
-                                            type: object
-                                          type: array
-                                          x-kubernetes-list-map-keys:
-                                            - name
-                                          x-kubernetes-list-type: map
-                                        limits:
-                                          additionalProperties:
-                                            anyOf:
-                                              - type: integer
-                                              - type: string
-                                            pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
-                                            x-kubernetes-int-or-string: true
-                                          type: object
-                                        requests:
-                                          additionalProperties:
-                                            anyOf:
-                                              - type: integer
-                                              - type: string
-                                            pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
-                                            x-kubernetes-int-or-string: true
-                                          type: object
-                                      type: object
                                     restartPolicy:
                                       type: string
                                     runtimeClassName:
@@ -3861,8 +3830,6 @@ spec:
                                         runAsUser:
                                           format: int64
                                           type: integer
-                                        seLinuxChangePolicy:
-                                          type: string
                                         seLinuxOptions:
                                           properties:
                                             level:
diff --git a/docs/api/rabbitmq.com.ref.asciidoc b/docs/api/rabbitmq.com.ref.asciidoc
index e9892478..de936e7d 100644
--- a/docs/api/rabbitmq.com.ref.asciidoc
+++ b/docs/api/rabbitmq.com.ref.asciidoc
@@ -86,6 +86,24 @@ More info: http://kubernetes.io/docs/user-guide/annotations
 |===
 
 
+[id="{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-overridetrimmed"]
+==== OverrideTrimmed 
+
+
+
+.Appears In:
+****
+- xref:{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-rabbitmqclusterspeccore[$$RabbitmqClusterSpecCore$$]
+****
+
+[cols="25a,75a", options="header"]
+|===
+| Field | Description
+| *`statefulSet`* __link:https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#rawextension-runtime-pkg[$$RawExtension$$]__ | Override configuration for the RabbitMQ StatefulSet.
+| *`service`* __xref:{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-service[$$Service$$]__ | Override configuration for the Service created to serve traffic to the cluster.
+|===
+
+
 [id="{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-persistentvolumeclaim"]
 ==== PersistentVolumeClaim 
 
@@ -263,7 +281,6 @@ Provides the ability to override the generated manifest of several child resourc
 .Appears In:
 ****
 - xref:{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-rabbitmqclusterspec[$$RabbitmqClusterSpec$$]
-- xref:{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-rabbitmqclusterspeccore[$$RabbitmqClusterSpecCore$$]
 ****
 
 [cols="25a,75a", options="header"]
@@ -458,6 +475,7 @@ Allows for the manifest of the created Service to be overwritten with custom con
 
 .Appears In:
 ****
+- xref:{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-overridetrimmed[$$OverrideTrimmed$$]
 - xref:{anchor_prefix}-github-com-rabbitmq-cluster-operator-v2-api-v1beta1-rabbitmqclusteroverridespec[$$RabbitmqClusterOverrideSpec$$]
 ****
 
-- 
2.51.0

